go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
+					oc.processObjectInTerminalState(objectsToRetry, obj, key, "add")
+					return
+				}
+
+				// If there is a delete entry with the same key, we got an add event for an object
+				// with the same name as a previous object that failed deletion.
+				// Destroy the old object before we add the new one.
+				if retryEntry := objectsToRetry.getObjRetryEntry(key); retryEntry != nil && retryEntry.oldObj != nil {
+					klog.Infof("Detected stale object during new object"+
+						" add of type %v with the same key: %s",
+						objectsToRetry.oType, key)
+					internalCacheEntry := oc.getInternalCacheEntry(objectsToRetry.oType, obj)
+					if err := oc.deleteResource(objectsToRetry, obj, internalCacheEntry); err != nil {
+						klog.Errorf("Failed to delete old object %s of type %v,"+
+							" during add event: %v", key, objectsToRetry.oType, err)
+						oc.recordErrorEvent(objectsToRetry.oType, obj, err)
+						objectsToRetry.unSkipRetryObj(key)
+						return
+					}
+					objectsToRetry.removeDeleteFromRetryObj(key)
+				}
+				start := time.Now()
+				if err := oc.addResource(objectsToRetry, obj, false); err != nil {
+					klog.Errorf("Failed to create %v object %s, error: %v",
+						objectsToRetry.oType, key, err)
+					oc.recordErrorEvent(objectsToRetry.oType, obj, err)
+					objectsToRetry.unSkipRetryObj(key)
+					return
+				}
+				klog.Infof("Creating %v %s took: %v", objectsToRetry.oType, key, time.Since(start))
+				objectsToRetry.deleteRetryObj(key, true)
+			},
+
+			UpdateFunc: func(old, newer interface{}) {
+				// skip the whole update if old and newer are equal
+				areEqual, err := areResourcesEqual(objectsToRetry.oType, old, newer)
+				if err != nil {
+					klog.Errorf("Could not compare old and newer resource objects of type %v: %v",
+						objectsToRetry.oType, err)
+					return
+				}
+				klog.V(5).Infof("Update event received for resource %s, old object is equal to new: %v",
+					objectsToRetry.oType, areEqual)
+				// QUESTION: should I do anything at all with the retry entries if the two objects are EQUAL?
+				// maybe set ignore=true?
+				if areEqual {
+					return
+				}
+
+				// get the object keys for newer and old (expected to be the same)
+				newKey, err := getResourceKey(objectsToRetry.oType, newer)
+				if err != nil {
+					klog.Errorf("Update of resource %v failed when looking up key of new obj: %v",
+						objectsToRetry.oType, err)
+					return
+				}
+				oldKey, err := getResourceKey(objectsToRetry.oType, old)
+				if err != nil {
+					klog.Errorf("Update of resource %v failed  when looking up key of old obj: %v",
+						objectsToRetry.oType, err)
+					return
+				}
+
+				// skip the whole update if the new object doesn't exist anymore in the API server
+				newer, err = oc.getResourceFromInformerCache(objectsToRetry.oType, newKey)
+				if err != nil && kerrors.IsNotFound(err) {
+					// QUESTION if not found in the API server, shouldn't we actually delete it?
```

Comment:
If it doens't exist in the kapi server anymore, then we should be getting a delete event soon for it so there is no need to process the delete here and then again for that event. It might be nice (and more proactive) to do it here in the future, but I dont think that is required for your PR. In the current watchPods code we do:

```
			pod, err := oc.watchFactory.GetPod(podNs, podName)
			if err != nil {
				klog.Warningf("Unable to get pod %s/%s for pod update, most likely it was already deleted",
					podNs, podName)
				return
			}
```

you need to handle if err!=nil && err != isNotFound